File: Messages.cpp
#include "Messages.hpp"

Messages* Messages::instance = new Messages;

Messages::Messages(QObject *parent) : QObject(parent)
{

}

Messages* Messages::qmlInstance(QQmlEngine*, QJSEngine*)
{
  return instance;
}

void Messages::message(QString text, QString details)
{
  instance->sendMessage(text, details);
}

void Messages::error(QString text, QString details)
{
  instance->sendError(text, details);
}

void Messages::sendMessage(QString text, QString details)
{
  history.append({text, details, false});
  emit gotMessage(text, details);
}

void Messages::sendError(QString text, QString details)
{
  history.append({text, details, true});
  emit gotError(text, details);
}

void Messages::reSendHistory()
{
  for (auto&& message : history) {
    if (message.isError) emit gotError(message.text, message.details);
    else emit gotMessage(message.text, message.details);
  }
}

File: src/messages.nim
{.used.}
import qt

var history: seq[tuple[text, details: string, isError: bool]]

var notifyMessage: proc(text, details: string) = proc(text, details: string) =
  history.add (text, details, false)

var notifyError: proc(text, details: string) = proc(text, details: string) =
  history.add (text, details, true)

proc sendMessage*(text: string, details = "") = notifyMessage(text, details)
proc sendError*(text: string, details = "") = notifyError(text, details)

type Messages = object

qobject Messages:
  proc message(text, details: string) {.signal.}
  proc error(text, details: string) {.signal.}

  proc init =
    notifyMessage = proc(text, details: string) = this.message(text, details)
    notifyError = proc(text, details: string) = this.error(text, details)

    for (text, details, isError) in history:
      if isError: this.error(text, details)
      else:       this.message(text, details)
    history = @[]

registerInQml Messages, "DMusic", 1, 0
